<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.31">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>MAGMA</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-e1a5c8363afafaef2c763b6775fbf3ca.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-7b1b06e26e1e2798ccde3208f86b5bbd.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar docked quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./output.html">Output</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">MAGMA</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./usage.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Usage</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./customizable-parameters.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Customizable Parameters</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./output.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Output</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#tutorials-and-presentations" id="toc-tutorials-and-presentations" class="nav-link active" data-scroll-target="#tutorials-and-presentations">Tutorials and Presentations</a></li>
  <li><a href="#interpretation" id="toc-interpretation" class="nav-link" data-scroll-target="#interpretation">Interpretation</a>
  <ul class="collapse">
  <li><a href="#qc-statistics-directory" id="toc-qc-statistics-directory" class="nav-link" data-scroll-target="#qc-statistics-directory">QC Statistics Directory</a>
  <ul class="collapse">
  <li><a href="#cohort" id="toc-cohort" class="nav-link" data-scroll-target="#cohort">cohort</a></li>
  <li><a href="#per_samplecoverage" id="toc-per_samplecoverage" class="nav-link" data-scroll-target="#per_samplecoverage"><strong>per_sample/coverage</strong></a></li>
  <li><a href="#per_samplemapping" id="toc-per_samplemapping" class="nav-link" data-scroll-target="#per_samplemapping"><strong>per_sample/mapping</strong></a></li>
  </ul></li>
  <li><a href="#analysis-directory" id="toc-analysis-directory" class="nav-link" data-scroll-target="#analysis-directory">Analysis Directory</a></li>
  <li><a href="#vcf_files-directory" id="toc-vcf_files-directory" class="nav-link" data-scroll-target="#vcf_files-directory"><code>vcf_files</code> Directory</a></li>
  <li><a href="#libraries-directory" id="toc-libraries-directory" class="nav-link" data-scroll-target="#libraries-directory">Libraries Directory</a></li>
  <li><a href="#samples-directory" id="toc-samples-directory" class="nav-link" data-scroll-target="#samples-directory">Samples Directory</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">MAGMA</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Here we briefly introduce the main outputs from MAGMA pipeline execution, please note that some outputs are optional and depends mainly on specific parameters to be generated.</p>
<ul>
<li><a href="#Interpretation">Interpretation</a></li>
</ul>
<section id="tutorials-and-presentations" class="level1">
<h1>Tutorials and Presentations</h1>
<p>Tim Huepink and Lennert Verboven created an in-depth tutorial of the features of the variant calling in MAGMA:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="https://www.youtube.com/watch?v=Kic2ItrJHj0"><img src="https://img.youtube.com/vi/Kic2ItrJHj0/maxresdefault.jpg" class="img-fluid figure-img"></a></p>
<figcaption>Video</figcaption>
</figure>
</div>
<p>We have also included a presentation (in PDF format) of the logic and workflow of the MAGMA pipeline as well as posters that have been presented at conferences. Please refer the <a href="./docs">docs</a> folder.</p>
</section>
<section id="interpretation" class="level1">
<h1>Interpretation</h1>
<p>The results directory produced by MAGMA is as follows:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/path/to/results_dir/</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="bu">.</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> QC_statistics</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> analyses</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="ex">└──</span> vcf_files</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="qc-statistics-directory" class="level2">
<h2 class="anchored" data-anchor-id="qc-statistics-directory">QC Statistics Directory</h2>
<p>In this directory you will find files related to the quality control carried out by the MAGMA pipeline. The structure is as follows:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/path/to/results_dir/QC_statistics</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> cohort</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="kw">|</span>   <span class="ex">└──</span> fastq_validation</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> └── multiqc</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>     └── multiqc_data</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="ex">└──</span> per_sample</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> coverage</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> fastqc</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> mapping</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="cohort" class="level3">
<h3 class="anchored" data-anchor-id="cohort">cohort</h3>
<p>Here you will find the <code>joint.merged_cohort_stats.tsv</code> which contains the QC statistics for all samples in the samplesheet and allows users to determine why certain samples failed to be incorporated in the cohort analysis steps</p>
<p>In addition, you’ll find the cohort-level MultiQC report generated by <code>per_sample/fastqc</code> analysis and the fastq validation report in <code>json</code> format.</p>
<section id="understanding-the-joint.merged_cohort_stats.tsv-file" class="level4">
<h4 class="anchored" data-anchor-id="understanding-the-joint.merged_cohort_stats.tsv-file">Understanding the joint.merged_cohort_stats.tsv file</h4>
<p>To accurately interpret the <code>joint.merged_cohort_stats.tsv</code> file and assess the quality of each sample analysis, we recommend that users consult the following table, which summarizes all relevant parameters.</p>
<table class="caption-top table">
<colgroup>
<col style="width: 30%">
<col style="width: 69%">
</colgroup>
<thead>
<tr class="header">
<th>Parameter</th>
<th>Meaning</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>SAMPLE</td>
<td>Identifier of the sample being analyzed.</td>
</tr>
<tr class="even">
<td>AVG_INSERT_SIZE</td>
<td>Average size (in base pairs) of the DNA fragment between paired-end reads.</td>
</tr>
<tr class="odd">
<td>MAPPED_PERCENTAGE</td>
<td>Percentage of reads successfully aligned to the reference genome.</td>
</tr>
<tr class="even">
<td>RAW_TOTAL_SEQS</td>
<td>Total number of raw reads sequenced, before any filtering or trimming.</td>
</tr>
<tr class="odd">
<td>AVERAGE_BASE_QUALITY</td>
<td>Average Phred quality score of bases across all reads.</td>
</tr>
<tr class="even">
<td>MEAN_COVERAGE</td>
<td>Average number of times each base of the reference is covered by aligned reads.</td>
</tr>
<tr class="odd">
<td>SD_COVERAGE</td>
<td>Standard deviation of the per-base coverage; indicates variation in coverage.</td>
</tr>
<tr class="even">
<td>MEDIAN_COVERAGE</td>
<td>Median coverage per base across the reference; less sensitive to outliers than the mean.</td>
</tr>
<tr class="odd">
<td>MAD_COVERAGE</td>
<td>Median Absolute Deviation of coverage; measures the variability around the median.</td>
</tr>
<tr class="even">
<td>PCT_EXC_ADAPTER</td>
<td>Percentage of bases or reads excluded due to presence of adapter sequences.</td>
</tr>
<tr class="odd">
<td>PCT_EXC_MAPQ</td>
<td>Percentage of bases or reads excluded due to low mapping quality (MAPQ).</td>
</tr>
<tr class="even">
<td>PCT_EXC_DUPE</td>
<td>Percentage of bases or reads marked as duplicates.</td>
</tr>
<tr class="odd">
<td>PCT_EXC_UNPAIRED</td>
<td>Percentage of reads excluded because they are unpaired in paired-end data.</td>
</tr>
<tr class="even">
<td>PCT_EXC_BASEQ</td>
<td>Percentage of bases excluded due to low base quality.</td>
</tr>
<tr class="odd">
<td>PCT_EXC_OVERLAP</td>
<td>Percentage of bases excluded due to overlapping paired-end reads.</td>
</tr>
<tr class="even">
<td>PCT_EXC_CAPPED</td>
<td>Percentage of bases excluded because they exceed the target regions.</td>
</tr>
<tr class="odd">
<td>PCT_EXC_TOTAL</td>
<td>Total percentage of excluded bases across all categories.</td>
</tr>
<tr class="even">
<td>PCT_1X</td>
<td>Percentage of target bases covered by at least 1 read.</td>
</tr>
<tr class="odd">
<td>PCT_5X</td>
<td>Percentage of target bases covered by at least 5 reads.</td>
</tr>
<tr class="even">
<td>PCT_10X</td>
<td>Percentage of target bases covered by at least 10 reads.</td>
</tr>
<tr class="odd">
<td>PCT_30X</td>
<td>Percentage of target bases covered by at least 30 reads.</td>
</tr>
<tr class="even">
<td>PCT_50X</td>
<td>Percentage of target bases covered by at least 50 reads.</td>
</tr>
<tr class="odd">
<td>PCT_100X</td>
<td>Percentage of target bases covered by at least 100 reads.</td>
</tr>
<tr class="even">
<td>LINEAGES FREQUENCIES</td>
<td>Relative abundance of different detected lineages.</td>
</tr>
<tr class="odd">
<td>MAPPED_NTM_FRACTION_16S</td>
<td>Fraction of 16S reads aligned to non-tuberculous mycobacteria (NTM).</td>
</tr>
<tr class="even">
<td>MAPPED_NTM_FRACTION_16S_THRESHOLD_MET</td>
<td>Whether the mapped NTM fraction in 16S exceeded the defined threshold (Boolean).</td>
</tr>
<tr class="odd">
<td>COVERAGE_THRESHOLD_MET</td>
<td>Whether the sample met the minimum average coverage requirement (Boolean).</td>
</tr>
<tr class="even">
<td>BREADTH_OF_COVERAGE_THRESHOLD_MET</td>
<td>Whether the required genome breadth was achieved (Boolean).</td>
</tr>
<tr class="odd">
<td>RELABUNDANCE_THRESHOLD_MET</td>
<td>Whether the relative abundance of target organism exceeded the required threshold (Boolean).</td>
</tr>
<tr class="even">
<td>ALL_THRESHOLDS_MET</td>
<td>Boolean indicating whether all quality and detection thresholds were met.</td>
</tr>
</tbody>
</table>
<p>The interpretation of median (middle value) or mean (average) coverage depth should be guided by the following categorization</p>
<table class="caption-top table">
<colgroup>
<col style="width: 23%">
<col style="width: 76%">
</colgroup>
<thead>
<tr class="header">
<th>Coverage Depth</th>
<th>Interpretation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>≥ 50X</td>
<td>A genome is analysable; nothing should be missed</td>
</tr>
<tr class="even">
<td>≥ 20X to &lt; 50X</td>
<td>A genome is analysable with nearly all minor variants detected</td>
</tr>
<tr class="odd">
<td>≥ 10X to &lt; 20X</td>
<td>A genome is analysable, but some minor variants may not be detected</td>
</tr>
<tr class="even">
<td>≥ 5X to &lt; 10X</td>
<td>A genome is analysable, but some variants may not be detected</td>
</tr>
<tr class="odd">
<td>&lt; 5X</td>
<td>A sample is not analysable</td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="per_samplecoverage" class="level3">
<h3 class="anchored" data-anchor-id="per_samplecoverage"><strong>per_sample/coverage</strong></h3>
<p>Contains the GATK WGSMetrics outputs for each of the samples in the samplesheet</p>
</section>
<section id="per_samplemapping" class="level3">
<h3 class="anchored" data-anchor-id="per_samplemapping"><strong>per_sample/mapping</strong></h3>
<blockquote class="blockquote">
<p>Contains the FlagStat and samtools stats for each of the samples in the samplesheet</p>
</blockquote>
</section>
</section>
<section id="analysis-directory" class="level2">
<h2 class="anchored" data-anchor-id="analysis-directory">Analysis Directory</h2>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/path/to/results_dir/analysis</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> cluster_analysis</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> drug_resistance</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> non-tuberculous_mycobacteria</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> phylogeny</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> spotyping</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="ex">└──</span> snp_distances</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><strong>Cluster Analysis</strong></li>
</ul>
<blockquote class="blockquote">
<p>Contains files related to clustering based on 5SNP and 12SNP cutoffs and inclunding and excluding complex regions <strong>.figtree files</strong>: These can be imported directly into Figtree for visualisation</p>
</blockquote>
<ul>
<li><strong>Drug Resistance</strong></li>
</ul>
<p>Organised based on the different types of variants as well as combined results:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/path/to/results_dir/analysis/drug_resistance</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> combined_resistance_summaries</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> combined_resistance_summaries_mixed_infection_samples</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> major_variants_xbs</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> minor_variants_lofreq</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> structural_variants_delly</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="ex">└──</span> tbprofiler_fastq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Each of the directories containing results related to the different variants (major | minor | structural) have text files that can be used to annotate the .treefiles produced by MAGMA in iToL (https://itol.embl.de)</p>
<p>The combined resistance results file contains a per-sample drug resistance summary based on the WHO Catalogue of <em>Mtb</em> mutations (https://www.who.int/publications/i/item/9789240082410)</p>
<p>MAGMA also notes the presence of all variants in in tier 1 and tier 2 drug resistance genes.</p>
<p>MAGMA will generated mixed infection reports and also optionally run tbprofiler from the fastq files for comparison purposes.</p>
<ul>
<li><strong>Non-Tuberculous Mycobacteria (NTM)</strong></li>
</ul>
<p>Contains a brief report of NTM presence on the submitted samples, in cohort and per_sample structure.</p>
<ul>
<li><strong>Phylogeny</strong></li>
</ul>
<p>Contains the outputs of the IQTree phylogenetic tree construction.</p>
<blockquote class="blockquote">
<p>:memo: By default we recommend that you use the <strong>ExDRIncComplex</strong> files as MAGMA was optimized to be able to accurately call positions on the edges of complex regions in the <em>Mtb</em> genome</p>
</blockquote>
<ul>
<li><strong>SNP distances</strong></li>
</ul>
<p>Contains the SNP distance tables in tsv format.</p>
<blockquote class="blockquote">
<p>:memo: By default we recommend that you use the <strong>ExDRIncComplex</strong> files as MAGMA was optimized to be able to accurately call positions on the edges of complex regions in the <em>Mtb</em> genome</p>
</blockquote>
<ul>
<li><strong>Spotyping</strong></li>
</ul>
<p>Contains a spoligotyping pattern prediction using <a href="https://github.com/xiaeryu/SpoTyping-v2.0/tree/master/SpoTyping-v2.0-commandLine">SpoTyping</a>.</p>
</section>
<section id="vcf_files-directory" class="level2">
<h2 class="anchored" data-anchor-id="vcf_files-directory"><code>vcf_files</code> Directory</h2>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/path/to/results_dir/vcf_files</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> cohort</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── combined_variant_files</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── minor_variants</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── multiple_alignment_files</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── raw_variant_files</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── snp_variant_files</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> └── structural_variants</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="ex">└──</span> per_sample</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> minor_variants</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> raw_variant_files</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> structural_variants</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><strong>Combined variant files</strong></li>
</ul>
<blockquote class="blockquote">
<p>Contains the cohort gvcfs based on major variants detected by the MAGMA pipeline</p>
</blockquote>
<ul>
<li><strong>Minor variants</strong></li>
</ul>
<blockquote class="blockquote">
<p>Merged vcfs of all samples, generated by LoFreq</p>
</blockquote>
<ul>
<li><strong>Multiple alignment files</strong></li>
</ul>
<blockquote class="blockquote">
<p>FASTA files for the generation of phylogenetic trees by IQTree</p>
</blockquote>
<ul>
<li><strong>Raw variant files</strong></li>
</ul>
<blockquote class="blockquote">
<p>Unfiltered indel and SNPs detected by the MAGMA pipeline</p>
</blockquote>
<ul>
<li><strong>SNP variant files</strong></li>
</ul>
<blockquote class="blockquote">
<p>Filtered SNPs detected by the MAGMA pipeline</p>
</blockquote>
<ul>
<li><strong>Structural variant files</strong></li>
</ul>
<blockquote class="blockquote">
<p>Unfiltered structural variants detected by the MAGMA pipeline</p>
</blockquote>
</section>
<section id="libraries-directory" class="level2">
<h2 class="anchored" data-anchor-id="libraries-directory">Libraries Directory</h2>
<blockquote class="blockquote">
<p>Contains files related to FASTQ validation and FASTQC analysis</p>
</blockquote>
</section>
<section id="samples-directory" class="level2">
<h2 class="anchored" data-anchor-id="samples-directory">Samples Directory</h2>
<blockquote class="blockquote">
<p>Contains vcf files for major|minor|structural variants for each individual samples</p>
</blockquote>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/TORCH-Consortium\.github\.io\/MAGMA\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>